### DISCUSSION TITLE
Couple of questions on writing a wrapper

### DISCUSSION THREAD

**Discussion by svengeance** [qpdf/qpdf#1100] [General]
Hey! Thanks for the amazing library - found it while looking for something that supports linearization. Definitely appreciate the completeness of it.

I'm working on [a C# wrapper over here](https://github.com/svengeance/QPdfSharp), as a small thank-you and exercise for myself (never done one before). I've got a couple of questions regarding on some expected usage patterns of QPDF that will help me make the it feel a bit more C#-natural.

## QPdf vs Jobs
While I will support both jobs regular qpdf object usage, I started with the latter. There's another C# library that solely supports json jobs, and my assumption was that, while jobs support more functionality, using the qpdf objects allows a bit more flexibility and perhaps performance.

- Is this assumption correct?
- Do you have any plans to expand the exposed C apis (e.g. for QPdfObjectHandle), or is the recommendation to write your own/use JSON?

## Memory & Writing
Reading the header documentation, I'm still unsure how best to handle writing multiple times from the same qpdf object. It doesn't seem like a typical usage pattern (the expected being load-modify-write), but I'm curious what your guidance is.

- Should writing multiple times be prevented?
- What happens to all of the internal buffers when writing to memory before cleanup? My testing shows the library is okay with writing to memory multiple times, but is that a support/anticipated use case is something like read, init_write, write, modify, write again?
  - I understand that after cleanup is called, all references to internal buffers should be invalidated. In C# it is common to dispose of objects at the end of a method, so it isn't unreasonable to expect many operations to occur on the qpdf object.

Thanks again for your work on this project, hopefully I can produce a wrapper that mirrors the beauty of it.

**Comment by jberkenbilt** [comment]
Re qpdf vs. jobs, your understanding is mostly correct. Everything in QPDFJobs is implemented in terms of other aspects of the qpdf API, so it's not strictly true that QPDFJobs is more functional, but more functionality is more easily exposed. I am actually designing yet another way to do page assembly that should address a number of shortcomings that have been in qpdf for years. I'm probably going to re-implement parts of QPDFJob in terms of that. I hope to create a discussion topic soon for that effort. I expect the assembly stuff to be accessible through QPDFJob and will probably not expose a C API for it.

There is an open issue somewhere about writing multiple times. My expectation is that it should work most of the time for most cases, but I have not made any guarantees in that API that it does. My intention is to sometime define the cases where it works and the cases where it doesn't, then bake that into the test suite and docs so that people can count on it. It's possible that that may be in scope for my current work, but I'm not sure yet. At this time, I'd say it's not generally safe to assume you can write the same qpdf object multiple times. I hope to change that. I have no intention of breaking cases that work. There are places where QPDFWriter modifies the QPDF in the writing process, so my analysis will have to be to study those cases and figure out when they matter and when they don't or to change QPDFWriter so that it never modifies the QPDF object. It's been a very long time since I've dug into the guts of that part of the code, so I just don't remember all the details.

The C API for QPDF and QPDFJob are both limited in comparison to the C++ API. There are various bindings in other languages that use the C API. I think there's a rust wrapper. I'm not sure whether the R wrapper uses the C API or C++. There's a Delphi wrapper that was the original reason I created the C API. In qpdf 11.7, I added a little bit of extra code to make it easier to connect other C++ code with the C API. The C API is not rich enough to expose everything that the C++ API can do. Just about anything involving direct manipulation of objects plus the full functionality of writing should be exposed, but more advanced operations like working with annotations or forms, converting pages into form XObjects, and other higher-level abstractions aren't exposed and probably never will be. The most complete foreign language bindings for qpdf would be pikepdf, which is in Python, and deeply integrates at the C++ level. I am not directly involved with pikepdf, but the pikepdf developer and I are in close collaboration. I even run pikepdf's test suite from qpdf's CI.

My time for qpdf is more limited than I wish it were, and right now, I'm finally diving into a feature I've had on the backlog for years. I will want to avoid being overly distracted from that since I have kept putting it off and letting other things sneak in front of it. That said, if you embark upon this project and run into things that are cumbersome, please let me know. I'm always interested in making improvements to the qpdf API that help with the kind of thing you're trying to do.

I've written this kind of fast....does this answer your questions? If not, let me know, and I can try to provide a little more detail. Thanks!

**Reply by svengeance** [reply]
No, it absolutely does. You put an amazing amount of effort into making this library accessible to both users and library authors like me, I very much appreciate it. Kudos to you. I don't expect I'll request anything from you to make authoring a wrapper easier, I'll work with what I've got and ask for your input on anything that's unclear.

RE PikePDF: Yeah I spent some time researching how they interop with the QPdfObjectHandle and they've indeed written an impressive amount of cpp to use with PyBind. I don't expect that I'll be able to put an equivalent amount of effort into this wrapper (not out of disinterest, simply time and other projects I want to do!), so I'll likely stick with the C APIs you've exposed.

RE Writing: I'll take all of this into consideration will design the library to be write-once. It's not _typical_, but I want to guarantee safety, and I don't want to do anything surprising like recreating qpdf objects after writing by holding onto backup references of the original. I imagine the overwhelming use-case is read->modify->write, so this shouldn't pose an issue. 

Again, thank you very much for your help. Do you mind if I leave this discussion open for any further questions? I'll do my best to keep them short and concise out of respect for your time. 

Cheers!

**Reply by jberkenbilt** [reply]
By all means, please feel free to leave this open. The issue I was looking for was #512. You can subscribe to that issue if you want to keep tabs on it. It is not currently prioritized. I'll also make a note there to remember to update this discussion if I do it. And thanks for your kind words about qpdf. It's been a labor of love for 20+ years, 15+ open source. It's very gratifying that people find it useful.

**Comment by svengeance** [comment]
@jberkenbilt Pleased to say that v1 is out. I've simply wrapped most of the QPdf document/pages API (so probably..60%? of qpdf-c. damn oh is a significant portion) so unfortunately the breadth of what it can do is a little limited, but it works out well enough. Over the next couple of months I'll work on the jobs API to facilitate more complex operations. I skipped out on the object handle manipulation for now since it's much more of a niche feature, but I may also implement that in time.

I don't really have anymore questions for now, but may in time with the jobs part. have a good weekend!🍻

**Reply by svengeance** [reply]
I've set C# up for success with this library by having a [daily GH action](https://github.com/svengeance/QPdfSharp/actions/workflows/build-publish-qpdf-runtime-libraries-nuget-on-schedule.yml) that checks for new versions of QPdf and uploads it to C#'s package manager (NuGet). Works out well if someone else wants to try their hand at a wrapper, or just use the binaries natively for whatever reason.

**Reply by jberkenbilt** [reply]
Great. Thanks. :-) I am going to be adding `QPDFAssembler` as a layer around page operations that will fall somewhere between the low-level API and `QPDFJob`. I haven't decided about a C API for it yet. It's issue #1104 if you're interested in considering that when the time comes.
